### ISSUE TITLE
[FR] Make fluent API thread safe

### ISSUE THREAD

**Issue by SaravananSathyanandhaQC** [mlflow/mlflow#9235]
### Willingness to contribute

Yes. I can contribute this feature independently.

### Proposal Summary

At the moment the fluent API is not threadsafe, as acknowledged in the existing documentation. I would like to propose making it thread safe.

### Motivation

> #### What is the use case for this feature?
We're running mlflow experiments kicked off from a web app, and thus want to run multiple experiments simultaneously.

> #### Why is this use case valuable to support for MLflow users in general?
Thread safety seems like a reasonable request in general to support, and the rest of Mlflow other than the Fluent API appears to be thread safe.

> #### Why is this use case valuable to support for your project(s) or organization?
Unblocks running multiple experiments simultaneously, rather than having to queue them up which would significantly impact the user's experience when there are multiple users on the web service.

> #### Why is it currently difficult to achieve this use case?
Due to IT constraints in our client environment where this web app is deployed we can't move experiment running to a separate process/cluster. We tried switching to the tracking API, but other than making our code a lot more verbose and messy, [mlflow.sklearn.log_model](https://mlflow.org/docs/latest/_modules/mlflow/sklearn.html#log_model) appears to only have a fluent API based implementation which is a blocker for us.

### Details

I'm not an expert on the mlflow library but it appears that changing these 3 lines - https://github.com/mlflow/mlflow/blob/master/mlflow/tracking/fluent.py#L72-L74

```
_active_run_stack = []
_active_experiment_id = None
_last_active_run_id = None
```

...to instead use [contextvars](https://docs.python.org/3/library/contextvars.html) might be sufficient to make it thread safe.

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by WeichenXu123** [comment]
This feature makes sense !

```
_active_run_stack = []
_active_experiment_id = None
_last_active_run_id = None
```

do these cover all global shared variables that causes thread-safe ?

**Comment by dbczumar** [comment]
What if we make the model logging APIs work with the existing thread-safe `MlflowClient` API? That sounds potentially easier than making the fluent API threadsafe.

**Comment by WeichenXu123** [comment]
Can we make `mlflow.autolog` emit warning that it is thread-unsafe when it detects it is called from different threads ?

**Comment by dbczumar** [comment]
> Can we make `mlflow.autolog` emit warning that it is thread-unsafe when it detects it is called from different threads ?

Maybe, but I'd prefer to support thread-safe model logging. Much more self-contained and also helpful for users of the `MlflowClient` API :D

**Comment by SaravananSathyanandhaQC** [comment]
> What if we make the model logging APIs work with the existing thread-safe `MlflowClient` API? That sounds potentially easier than making the fluent API threadsafe.

Well I think the making fluent API threadsafe is “easy” if my assessment is right that it’s only those 3 global variables that are the issue.

I can look at the model logging to make it support using Mlflow client too, probably needs an optional arg for run id or a separate method that matches mlflow tracking API closer. But I do think the fluent API makes code so much simpler (not having to pass run IDs everywhere) that it would be nice for it to be thread safe as well.

**Comment by SaravananSathyanandhaQC** [comment]
Started a PR above for making fluent API threadsafe https://github.com/mlflow/mlflow/pull/9249 - will separately look at making model logging using MlflowClient. I think both changes are valuable independently of each other.

**Comment by FlorianThaler** [comment]
Hello there.

I ran into a similar issue. I cannot use properly mlflow functionalities in a multi-threaded application. Are there any news regarding thread-safety of mlflow?

BR, Flo

**Comment by Galileo-Galilei** [comment]
For anyone looking for it, it seems to be fixed for ``mlflow>2.18.0`` by https://github.com/mlflow/mlflow/pull/13456

**Comment by mazer-ai** [comment]
Just for the record -- I don't think the Client API is fully thread safe. Accessing the artifact store using the Client functions drops back into the fluent API and depends on the global state set with `mlflow.set_tracking_uri()`. This means you can't access all functionality in the client object w/o potentially changing global state and impacting other threads and client objects.

See https://github.com/mlflow/mlflow/issues/14931

**Comment by B-Step62** [comment]
This is supported as of MLflow 3. Please file a new issue when you encounter an issue related to thread safety when using fluent APIs.
